src/app/clusters/temperature-control-server/temperature-control-server.cpp

```diff
@@ -0,0 +1,161 @@
+        TemperatureControl::GetSupportedTemperatureLevelsManager();
+
+    if (TemperatureControl::Attributes::SupportedTemperatureLevels::Id == aPath.mAttributeId)
+    {
+        const TemperatureControl::SupportedTemperatureLevelsManager::TemperatureLevelOptionsProvider
+            temperatureLevelOptionsProvider =
+                gSupportedTemperatureLevelManager->GetTemperatureLevelOptionsProvider(aPath.mEndpointId);
+        if (temperatureLevelOptionsProvider.begin() == nullptr)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+        CHIP_ERROR err;
+        err = aEncoder.EncodeList([temperatureLevelOptionsProvider](const auto & encoder) -> CHIP_ERROR {
+            const auto * end = temperatureLevelOptionsProvider.end();
+            for (auto * it = temperatureLevelOptionsProvider.begin(); it != end; ++it)
+            {
+                auto & temperatureLevelOption = *it;
+                ReturnErrorOnFailure(encoder.Encode(temperatureLevelOption));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool TemperatureControlHasFeature(EndpointId endpoint, TemperatureControl::Feature feature)
+{
+    bool success;
+    uint32_t featureMap;
+    success = (Attributes::FeatureMap::Get(endpoint, &featureMap) == EMBER_ZCL_STATUS_SUCCESS);
+
+    return success ? ((featureMap & to_underlying(feature)) != 0) : false;
+}
+
+/**********************************************************
+ * Callbacks Implementation
+ *********************************************************/
+
+bool emberAfTemperatureControlClusterSetTemperatureCallback(app::CommandHandler * commandObj,
+                                                            const app::ConcreteCommandPath & commandPath,
+                                                            const Commands::SetTemperature::DecodableType & commandData)
+{
+    auto & targetTemperature      = commandData.targetTemperature;
+    auto & targetTemperatureLevel = commandData.targetTemperatureLevel;
+    EndpointId endpoint           = commandPath.mEndpointId;
+    Status status                 = Status::Success;
+
+    if (TemperatureControlHasFeature(endpoint, TemperatureControl::Feature::kTemperatureNumber))
+    {
+        if (!targetTemperature.HasValue())
+        {
+            status = Status::Failure;
+        }
+
+        int16_t minTemperature = 0;
+        int16_t maxTemperature = 0;
+        if (MinTemperature::Get(endpoint, &minTemperature) != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            status = Status::Failure;
+        }
+        if (MaxTemperature::Get(endpoint, &maxTemperature) != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            status = Status::Failure;
+        }
+
+        if (targetTemperature.Value() < minTemperature || targetTemperature.Value() > maxTemperature)
```

Comment:
This is going to crash if `!targetTemperature.HasValue()`.  The overall error-handling logic of this whole block does not really make sense.

Reply:
Updated check